src/errors/kinds.rs

```diff
@@ -508,3 +514,75 @@ impl ErrorKind {
         }
     }
 }
+
+#[pyclass(extends=PyValueError, module="pydantic_core._pydantic_core")]
+#[derive(Debug, Clone)]
+pub struct PydanticErrorKind {
+    kind: String,
+    message_template: String,
+    context: Option<Py<PyDict>>,
+}
+
+#[pymethods]
+impl PydanticErrorKind {
+    #[new]
+    pub fn py_new(py: Python, kind: String, message_template: String, context: Option<&PyDict>) -> Self {
+        Self {
+            kind,
+            message_template,
+            context: context.map(|c| c.into_py(py)),
+        }
+    }
+
+    #[getter]
+    pub fn kind(&self) -> String {
+        self.kind.clone()
+    }
+
+    #[getter]
+    pub fn message_template(&self) -> String {
```

Comment:
This part I am struggling to understand what is needed. Does this mean there would be a function like this (taking `message_template` as an example):

```rust
    #[getter]
    pub fn message_template(&self, py: Python) -> PyResult<String> {
        ErrorKind::PydanticError {
            value_error: String::from("Pydantic error"),
        }
        .render_message(py)
    }
```

And in `render_message` from `ErrorKind` add a line like:

```rust
Self::PydanticError { value_error } => render!(self, value_error),
```

I don't think this is correct, but I'm not sure what to do instead.

Reply:
There are two cases:
* we're returning the error within pydantic-core validation logic, here there's no need for any of this, we just look up the matching `ErrorKind`, fill it's attributes from `context`, this will need another long match statement
* python methods like `message_context`, which won't actually be required during validation, but might help users understand what's going on. Probably easiest to use the above `(kind: &str, context) -> ErrorKind` logic to get the error kind, then return it's message template

Please have a crack at that, if you're still struggling, it might be easier for me to push to your PR to explain what I mean?

Reply:
I don't see how `PydanticErrorKind` gets enough information to know which error type is the correct error. If you want to push to my branch that would be good with me. I don't want to continue to waste your time with my inability to comprehend the requirements.

Reply:
Ok, will do.

Sorry if I wasn't clear enough. `PydanticErrorKind` takes `kind` as an argument, that kind must match one of the `ErrorKind.kind()` strings.

Reply:
done.